File: api-fastify/src/auth/keycloak.js
import _ from 'lodash';
import config from 'config';
import TTLCache from '@isaacs/ttlcache';
import FastifyOauth2 from '@fastify/oauth2';

export default async (app) => {
  const urlOpenidConfiguration = `${config.get('keycloak.issuer')}/.well-known/openid-configuration`;
  const { end_session_endpoint, userinfo_endpoint, token_endpoint, revocation_endpoint } = await fetch(urlOpenidConfiguration).then(r => r.json());
  app.log.info(urlOpenidConfiguration);

  app.register(FastifyOauth2, {
    name: 'keycloak',
    credentials: {
      client: config.get('keycloak.client'),
    },
    discovery: {
      issuer: config.get('keycloak.issuer'),
    },
    startRedirectPath: '/login',
    callbackUri: config.get('keycloak.urls.callback'),
    scope: config.get('keycloak.scope'),
  });

  const tokenCache = new TTLCache({ ttl: 60000 });

  app.get('/callback', async (req, res) => {
    // app.log.info(req.query);
    const { token } = await app.keycloak.getAccessTokenFromAuthorizationCodeFlow(req);
    // app.log.info({ token });
    const { session_state } = token;
    tokenCache.set(session_state, token);
    const search = new URLSearchParams({ session_state });
    res.redirect(`${config.get('keycloak.urls.callbackRedirect')}?${search.toString()}`);
  });

  app.post('/tokens', async (req, res) => {
    res.send(tokenCache.get(req.body.session_state));
    tokenCache.delete(req.body.session_state);
  });

  app.get('/userinfo', (req, res) => {
    fetch(userinfo_endpoint, { headers: _.pick(req.headers, 'authorization') }).then((resp) => {
      res.status(resp.status).type('application/json').send(resp.body);
    });
  });

  app.get('/logout', (req, res) => {
    const url = new URL(end_session_endpoint);
    url.search = new URLSearchParams(req.query).toString();
    res.redirect(url.href);
  });

  app.post('/tokens/refresh', (req, res) => {
    fetch(token_endpoint, {
      method: 'POST',
      body: new URLSearchParams({
        client_id: config.get('keycloak.client.id'),
        client_secret: config.get('keycloak.client.secret'),
        refresh_token: req.body.refresh_token,
        grant_type: 'refresh_token',
      }),
    }).then((resp) => {
      res.status(resp.status).type('application/json').send(resp.body);
    });
  });

  app.post('/tokens/revoke', (req, res) => {
    fetch(revocation_endpoint, {
      method: 'POST',
      body: new URLSearchParams({
        client_id: config.get('keycloak.client.id'),
        client_secret: config.get('keycloak.client.secret'),
        ...req.body,
      }),
    }).then((resp) => {
      res.status(resp.status).type('application/json').send(resp.body);
    });
  });

  app.post('/tokens/exchange', (req, res) => {
    app.log.info({
      client_id: config.get('keycloak.client.id'),
      client_secret: config.get('keycloak.client.secret'),
      grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
      subject_token: req.body.access_token,
      audience: 'apidemo',
      // scope: 'openid',
    });
    fetch(token_endpoint, {
      method: 'POST',
      body: new URLSearchParams({
        client_id: config.get('keycloak.client.id'),
        client_secret: config.get('keycloak.client.secret'),
        grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
        subject_token: req.body.access_token,
        audience: 'apidemo',
        scope: 'openid',
      }),
    }).then((resp) => {
      res.status(resp.status).type('application/json').send(resp.body);
    });
  });

  app.get('/configurations', async () => {
    const configurations = await fetch(urlOpenidConfiguration).then(r => r.json());
    return {
      client_id: config.get('keycloak.client.id'),
      ...configurations,
    };
  });

};

export const autoPrefix = '/keycloak';

File: api-fastapi/src/auth/keycloak.py
import httpx
import urllib.parse
from cachetools import TTLCache
from fastapi import APIRouter, Request, Response, Header
from fastapi.responses import RedirectResponse
from authlib.integrations.starlette_client import OAuth

from config import settings

router = APIRouter(prefix="/keycloak")

oauth = OAuth()

tokenCache = TTLCache(maxsize=256, ttl=60)

oauth.register(
    name="keycloak",
    server_metadata_url=f'{settings.keycloak_openid_host}/realms/{settings.keycloak_openid_realm}/.well-known/openid-configuration',
    client_id=settings.keycloak_client_id,
    client_secret=settings.keycloak_client_secret,
    client_kwargs={"scope": settings.keycloak_openid_scope},
)

@router.get('/configurations')
async def configurations(req: Request):
    return {
        ** req.app.state.configurations,
        'client_id': settings.keycloak_client_id,
    }

@router.get("/callback", name="auth_callback")
async def auth_callback(request: Request, session_state: str):
    tokens = await oauth.keycloak.authorize_access_token(request)
    tokenCache[session_state] = tokens
    query = urllib.parse.urlencode({ 'session_state': session_state })
    return RedirectResponse(f'{settings.url_frontend}/#/keycloak?{query}')

@router.post("/tokens")
async def auth_tokens(request: Request):
    return tokenCache.pop((await request.json())['session_state'], None)

@router.post("/tokens/refresh")
async def auth_tokens_refresh(request: Request):
    async with httpx.AsyncClient() as client:
        resp = await client.post(f'{request.app.state.configurations['token_endpoint']}', data = {
            'client_id': settings.keycloak_client_id,
            'client_secret': settings.keycloak_client_secret,
            'refresh_token': (await request.json())['refresh_token'],
            'grant_type': 'refresh_token',
        })
    return Response(status_code=resp.status_code, content=resp.content)

@router.post("/tokens/revoke")
async def auth_tokens_revoke(request: Request):
    async with httpx.AsyncClient() as client:
        resp = await client.post(f'{request.app.state.configurations['revocation_endpoint']}', data = {
            'client_id': settings.keycloak_client_id,
            'client_secret': settings.keycloak_client_secret,
            **(await request.json())
        })
    return Response(status_code=resp.status_code, content=resp.content)

@router.get("/login")
async def auth_login(request: Request):
    redirect_uri = f'{settings.url_backend}/auth/keycloak/callback'
    return await oauth.keycloak.authorize_redirect(request, redirect_uri)

@router.get("/logout")
def auth_logout(request: Request):
    return RedirectResponse(f'{request.app.state.configurations['end_session_endpoint']}?{request.query_params}')

@router.get("/userinfo")
async def auth_userinfo(request: Request, authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        resp = await client.get(f'{request.app.state.configurations['userinfo_endpoint']}', headers={ 'authorization': authorization })
    return Response(status_code=resp.status_code, content=resp.content)
